src/main/java/org/opensearch/neuralsearch/query/HybridQueryDocIdStream.java

```diff
@@ -52,33 +73,22 @@ public void forEach(int upTo, CheckedIntConsumer<IOException> consumer) throws I
                     hybridBulkScorer.getHybridSubQueryScorer().getSubQueryScores()[subQueryIndex] = scoreOfDocIdForSubQuery;
                 }
                 // process the document with its base offset
-                int doc = base | docIndexInWindow;
-                if (doc < upTo + base) {
-                    consumer.accept(doc);
-                    this.upTo++;
-                }
+                consumer.accept(docId);
                 // reset scores after processing of one doc, this is required because scorer object is re-used
                 hybridBulkScorer.getHybridSubQueryScorer().resetScores();
+                // clear bit from the local bitset copy to indicate that it has been consumed
+                matchingBitSet.clear(docIndexInWindow);
                 // reset bit for this doc id to indicate that it has been consumed
                 bits ^= 1L << numberOfTrailingZeros;
             }
         }
     }
 
-    @Override
-    public int count(int upTo) throws IOException {
-        int[] count = new int[1];
-        forEach(upTo, (doc -> count[0]++));
-        return count[0];
-    }
-
-    @Override
-    public boolean mayHaveRemaining() {
-        return this.upTo + 1 < hybridBulkScorer.getMaxDoc();
-    }
-
-    public void setBase(int base) {
-        this.base = base;
-        this.upTo = base;
+    // This lazy loading is necessary because the matching bit isn't available at the time this class is constructed.
+    private FixedBitSet getLocalMatchingBitSet() {
+        if (localMatchingBitSet == null) {
+            localMatchingBitSet = hybridBulkScorer.getMatching().clone();
```

Comment:
Shall we verify behavior with documents trigger multiple windows? Current tests appear to use smaller datasets where this scenario wouldn't occur.

Reply:
Not sure if such case should be implemented as integ test. Shouldn't it be captured by benchmark test or micro benchmark test instead which is out of scope for this PR?

Reply:
Added a unit test instead.